### PR TITLE
ARROW-13252: [C++] Add offset to CSV error reporting

### PR DESCRIPTION
**Pull request by n3world**
Add the offset of the start of a row to parsing and decoding errors.

In order to know the row offset during decoding a new array of offsets was added to DataBatch which indicate the offset of the row relative to the offset of the batch. This adds an overhead of an additional vector and an array of ints with one int per row

### CODE REVIEW

**Comment by n3world** [comment]
@pitrou or @westonpace any opinions on this MR?

File: cpp/src/arrow/csv/reader.cc [hunk 1]
```diff
@@ -391,17 +391,20 @@ namespace {
...
         count_rows_(first_row >= 0),
-        num_rows_seen_(first_row) {}
+        num_rows_seen_(first_row),
+        next_block_offset_(offset) {}
 
   Result<ParsedBlock> operator()(const CSVBlock& block) {
     constexpr int32_t max_num_rows = std::numeric_limits<int32_t>::max();
-    auto parser = std::make_shared<BlockParser>(
-        io_context_.pool(), parse_options_, num_csv_cols_, num_rows_seen_, max_num_rows);
+    next_block_offset_ += block.bytes_skipped;
```
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:404]: At the moment streaming parsing is always serial but that is due to a bug/limitation (ARROW-13155).  Eventually, `operator()` will be called multiple times at once (across multiple blocks).  I think this breaks both `num_rows_seen_` and `next_block_offset_` but we can fix it then.

File: cpp/src/arrow/csv/reader.cc [hunk 2]
```diff
@@ -742,6 +751,25 @@ class ReaderMixin {
     return ParseResult{std::move(parser), static_cast<int64_t>(parsed_size)};
   }
 
+  /**
+   * @brief add the `size()` of all buffers in `block` to `parser_offset_`
+   * @return the parser offset which should be used by the caller
+   */
+  int64_t UpdateParserOffset(const CSVBlock& block) {
+    parser_offset_ += block.bytes_skipped;
+    auto result = parser_offset_;
```
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:760]: It's been a hard habit for me to break but `result` is not a good variable name.  Maybe `updated_offset` or something like that?

File: cpp/src/arrow/csv/reader.cc [hunk 3]
```diff
@@ -742,6 +751,25 @@ class ReaderMixin {
     return ParseResult{std::move(parser), static_cast<int64_t>(parsed_size)};
   }
 
+  /**
+   * @brief add the `size()` of all buffers in `block` to `parser_offset_`
+   * @return the parser offset which should be used by the caller
```
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:756]: Use `\brief` and `\return` and `///` comments

**Review by westonpace — CHANGES_REQUESTED** [review]
Sorry for the wait for feedback.  Have you run the parsing benchmarks in `parser_benchmark.cc`?  Does capturing the offset have any noticeable effect on performance?

File: cpp/src/arrow/csv/parser.cc [hunk 4]
```diff
@@ -100,6 +100,47 @@ class PresizedDataWriter {
...
+  explicit TypedBuffer(MemoryPool* pool, int64_t capacity)
+      : size_(0), capacity_(capacity), mark_(-1) {
+    buffer_ = *AllocateResizableBuffer(capacity_ * sizeof(Type), pool);
+    typed_buffer_ = reinterpret_cast<Type*>(buffer_->mutable_data());
+  }
+
+  void Append(const Type& value) {
+    if (ARROW_PREDICT_FALSE(size_ == capacity_)) {
+      capacity_ *= 2;
+      ARROW_CHECK_OK(buffer_->Resize(capacity_ * sizeof(Type)));
```
[the remark this answers is not in the document]
Comment (n3world) [reply cpp/src/arrow/csv/parser.cc:117]: As noted before this is a void method, so I cannot return the status here.

**Comment by n3world** [comment]
> Sorry for the wait for feedback. Have you run the parsing benchmarks in `parser_benchmark.cc`? Does capturing the offset have any noticeable effect on performance?

I was getting some wide variations between runs but these are the best numbers I got for master and this branch
master:

> ----------------------------------------------------------------------------------
Benchmark                        Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------------
ChunkCSVQuotedBlock         168985 ns       168981 ns         4117 bytes_per_second=959.423M/s
ChunkCSVEscapedBlock        155557 ns       155555 ns         4460 bytes_per_second=980.924M/s
ChunkCSVNoNewlinesBlock        147 ns          147 ns      4741333 bytes_per_second=0/s
ParseCSVQuotedBlock         263473 ns       263469 ns         2646 bytes_per_second=615.346M/s
ParseCSVEscapedBlock        209135 ns       209132 ns         3362 bytes_per_second=729.624M/s
ParseCSVFlightsExample     2175256 ns      2175243 ns          320 bytes_per_second=446.642M/s
ParseCSVVehiclesExample   15967256 ns     15967111 ns           44 bytes_per_second=718.222M/s
ParseCSVStocksExample      3463566 ns      3463298 ns          203 bytes_per_second=605.926M/s

This branch:
> ----------------------------------------------------------------------------------
Benchmark                        Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------------
ChunkCSVQuotedBlock         169009 ns       169006 ns         4093 bytes_per_second=959.283M/s
ChunkCSVEscapedBlock        156445 ns       156443 ns         4467 bytes_per_second=975.356M/s
ChunkCSVNoNewlinesBlock        149 ns          149 ns      4749759 bytes_per_second=0/s
ParseCSVQuotedBlock         369561 ns       369551 ns         1882 bytes_per_second=438.707M/s
ParseCSVEscapedBlock        367681 ns       367671 ns         1867 bytes_per_second=415.012M/s
ParseCSVFlightsExample     2538161 ns      2538102 ns          278 bytes_per_second=382.788M/s
ParseCSVVehiclesExample   16641194 ns     16639585 ns           42 bytes_per_second=689.196M/s
ParseCSVStocksExample      3119450 ns      3119364 ns          224 bytes_per_second=672.734M/s

No significant difference that I can see

**Comment by pitrou** [comment]
> Most have no significant difference but ParseCSVQuotedBlock, ParseCSVEscapedBlock and ParseCSVFlightsExample did show performance degradation

Sorry for not answering earlier; but the degradation looks quite concerning indeed. A minor difference (e.g. 3-4%) would be ok, but we seem to be getting a 30% slowdown in some of these cases.

**Comment by n3world** [comment]
I think I figured out the issue. It was the overhead of checking the buffer bounds on every append. I changed it back to not checking bounds for the fixed size buffer and numbers look much better

```
----------------------------------------------------------------------------------
Benchmark                        Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------------
ChunkCSVQuotedBlock         174105 ns       174102 ns         4012 bytes_per_second=931.204M/s
ChunkCSVEscapedBlock        143420 ns       143417 ns         4863 bytes_per_second=1063.95M/s
ChunkCSVNoNewlinesBlock        179 ns          179 ns      4276522 bytes_per_second=0/s
ParseCSVQuotedBlock         286099 ns       286084 ns         2434 bytes_per_second=566.703M/s
ParseCSVEscapedBlock        267409 ns       267403 ns         2612 bytes_per_second=570.629M/s
ParseCSVFlightsExample     2238658 ns      2238612 ns          310 bytes_per_second=433.999M/s
ParseCSVVehiclesExample   16297590 ns     16297154 ns           43 bytes_per_second=703.677M/s
ParseCSVStocksExample      3236563 ns      3236484 ns          216 bytes_per_second=648.39M/s
```

**Comment by n3world** [comment]
Any opinion on this @pitrou or @westonpace ?

**Comment by pitrou** [comment]
While the micro-benchmarks don't show any regression, it's quite conceivable that a more memory-heavy workload would show a regression. After all, we're building up an additional buffer, even though it's smaller than the values buffer. So I'm a bit lukewarm on this, especially as the motivation is a rather niche use case.

**Comment by n3world** [comment]
> While the micro-benchmarks don't show any regression, it's quite conceivable that a more memory-heavy workload would show a regression.

That could be an issue if the system is memory constrained relative to the size of the csv

> So I'm a bit lukewarm on this, especially as the motivation is a rather niche use case.

I'm not sure how returning the offset of an error is a niche use case especially when using threaded parsing the line number is not known so when an error occurs there is no line number to indicate where the error was encountered

**Comment by pitrou** [comment]
> That could be an issue if the system is memory constrained relative to the size of the csv

More generally though, more allocations and more memory writing can increase the impact on the CPU cache and memory subsystem, even if lots of RAM are still available.

**Comment by n3world** [comment]
It seems like this functionality is not desired. Should I just close the PR and associated jira ticket?

**Comment by pitrou** [comment]
@n3world Sorry. Judging by the amount of complexity and runtime overhead it adds, I think it's not desirable indeed.

**Pull request closed without merging**